neuronsimulator/nrn
# Porting CoreNEURON profiler wrappers to NEURON

This PR attempts to address #1421 

After trying several options to include CoreNEURON's implementation directly in NEURON I resorted to port the code over for now:

- Moving the code to NRN and calling it from CoreNEURON is not a good solution because GPU related stuff will still have to be factored out and kept in CoreNEURON.
- Using the CoreNEURON code from NRN means either starting to include crnrn headers when building nrn, which is something we don't do currently or doing dynamic library calls (similar to how we call the coreNEURON solver from neuron). This strikes me as a very expensive way to call profilers and seems overly complex.

## How to test

```
mkdir build && cd build
cmake .. -DNRN_ENABLE_PROFILING=ON -DNRN_PROFILER=caliper -DNRN_ENABLE_TESTS=ON
make
cd test/testcorenrn_bbcore/neuron
CALI_CONFIG=runtime-report,calc.inclusive NRNHOME=/Users/awile/projects/cellular/nrn/build2 NEURONHOME=/Users/awile/projects/cellular/nrn/build2/share/nrn /Users/awile/projects/cellular/nrn/build2/bin/nrniv testbbcore.hoc
```

sample output:

```
NEURON -- VERSION 8.0a-632-g12e885a89+ enh/1421 (12e885a89+) 2021-09-01
Duke, Yale, and the BlueBrain Project -- Copyright 1984-2021
See http://neuron.yale.edu/neuron/credits

loading membrane mechanisms from x86_64/.libs/libnrnmech.so
Additional mechanisms from files
 "./Gfluct3.mod" "./hhderiv.mod" "./hhkin.mod" "./hhwatch.mod" "./nacum.mod" "./vecevent.mod"
arg_tstop  10
arg_coreneuron  0
arg_coreneuron_filemode  0
arg_dump_coreneuron_model  0
arg_coreneuron_gpu  0
arg_pattern
arg_psolve_test_mode  0
	1
	1
	0
	1
	0
	0
	1
	10
psolve time 0.0300002
	0
Path                  Time (E) Time (I) Time % (E) Time % (I)
psolve                0.001221 0.018629   6.244885  95.279255
  timestep            0.008836 0.017408  45.192308  89.034370
    state-update      0.001361 0.001361   6.960925   6.960925
    update            0.001343 0.001343   6.868863   6.868863
    second_order_cur  0.001413 0.001413   7.226882   7.226882
    matrix-solver     0.001414 0.001414   7.231997   7.231997
    setup_tree_matrix 0.001480 0.001480   7.569558   7.569558
    deliver_events    0.001561 0.001561   7.983838   7.983838
```

## What's left to do

- [x] add timers for channels and other solver code paths
- [x] Add a caliper benchmark into tests
- [x] add into one of the CI plans caliper and a benchmark test
